### PR TITLE
[NEW][8.0] portal_welcome_email_template

### PR DESCRIPTION
**Pull request by sergio-teruel**
Customize portal access email message with a email template.

### CODE REVIEW

File: portal_welcome_email_template/wizards/portal_wizard.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    """
+    _inherit = 'portal.wizard.user'
+
+    @api.model
+    def _send_email(self, wizard_user):
+        """ send notification email to a new portal user
+            @param wizard_user: browse record of model portal.wizard.user
+            @return: the id of the created mail.mail record
+        """
+        this_context = self._context
```
Comment (hbrunn) [inline portal_welcome_email_template/wizards/portal_wizard.py]: use `self.env.context`

File: portal_welcome_email_template/wizards/portal_wizard.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        this_user = self.env['res.users'].sudo().browse(self._uid)
```
Comment (hbrunn) [inline portal_welcome_email_template/wizards/portal_wizard.py]: use `self.env.user.sudo()` (and only sudo if really necessary)

File: portal_welcome_email_template/wizards/portal_wizard.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+# (c) 2015 Incaser Informatica S.L. - Carlos Dauden
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+from openerp import models, api
+from openerp import exceptions
+from openerp.tools.translate import _
+
+
+class WizardUser(models.TransientModel):
+    """
```
Comment (pedrobaeza) [inline portal_welcome_email_template/wizards/portal_wizard.py]: Why adding again this docstring?

File: portal_welcome_email_template/wizards/portal_wizard.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    @api.model
+    def _send_email(self, wizard_user):
+        """ send notification email to a new portal user
+            @param wizard_user: browse record of model portal.wizard.user
+            @return: the id of the created mail.mail record
+        """
+        this_context = self._context
+        this_user = self.env['res.users'].sudo().browse(self._uid)
+        if not this_user.email:
```
Comment (hbrunn) [inline portal_welcome_email_template/wizards/portal_wizard.py]: the template determines the from: header, so why this check?

File: portal_welcome_email_template/wizards/portal_wizard.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            raise exceptions.Warning(
+                _('Email Required'),
+                _('You must have an email address in your User Preferences to '
+                  'send emails.'))
+
+        # determine subject and body in the portal user's language
+        user = self.sudo()._retrieve_user(wizard_user)
```
Comment (hbrunn) [inline portal_welcome_email_template/wizards/portal_wizard.py]: why is sudo necessary here?

File: portal_welcome_email_template/wizards/portal_wizard.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        context = dict(this_context or {}, lang=user.lang)
+        ctx_portal_url = dict(context, signup_force_type_in_url='')
+        portal_url = user.partner_id.with_context(
```
Comment (hbrunn) [inline portal_welcome_email_template/wizards/portal_wizard.py:16]: you don't need to replace the whole context, `.with_context(lang=user.lang, signup_force_type_in_url='')` would suffice here

File: portal_welcome_email_template/wizards/portal_wizard.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            ctx_portal_url)._get_signup_url_for_action()[user.partner_id.id]
+        user.partner_id.with_context(context).signup_prepare()
```
Comment (hbrunn) [inline portal_welcome_email_template/wizards/portal_wizard.py]: se above

File: portal_welcome_email_template/wizards/portal_wizard.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+        template = self.env.ref(
+            'portal_welcome_email_template.portal_welcome_email')
+        ctx = this_context.copy()
+        ctx.update({
+            'login': user.login,
+            'portal_url': portal_url,
+            'db': self._cr.dbname,
```
Comment (hbrunn) [inline portal_welcome_email_template/wizards/portal_wizard.py]: use `self.env.cr`

File: portal_welcome_email_template/wizards/portal_wizard.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+            'portal': wizard_user.wizard_id.portal_id.name,
+            'signup_url': user.signup_url,
+            'welcome_message': wizard_user.wizard_id.welcome_message or "",
+        })
+        result = template.with_context(ctx).generate_email_batch(
+            template.id, [user.id])[user.id]
+        return self.env['mail.mail'].with_context(this_context).create(result)
```
Comment (hbrunn) [inline portal_welcome_email_template/wizards/portal_wizard.py]: `this_context` is the current context, no need to set it again

File: portal_welcome_email_template/wizards/portal_wizard.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        ctx = this_context.copy()
+        ctx.update({
+            'login': user.login,
+            'portal_url': portal_url,
+            'db': self._cr.dbname,
+            'portal': wizard_user.wizard_id.portal_id.name,
+            'signup_url': user.signup_url,
+            'welcome_message': wizard_user.wizard_id.welcome_message or "",
+        })
+        result = template.with_context(ctx).generate_email_batch(
```
Comment (hbrunn) [inline portal_welcome_email_template/wizards/portal_wizard.py]: why not `template.with_context(ctx).send_mail(user.id)`: https://github.com/OCA/OCB/blob/8.0/addons/email_template/email_template.py#L543 ?

**Comment by hbrunn** [comment]
please add tests for your python code

File: portal_welcome_email_template/README.rst [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+=============================
+Portal Welcome Email Template
+=============================
+
+This module extends the functionality of portal module to allow customize the
+email message sent to customer in email template section.
+The email sent is sent to queue.
```
Comment (elicoidal) [inline portal_welcome_email_template/README.rst]: The sent email is sent to queue.

File: portal_welcome_email_template/__openerp__.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+# (c) 2015 Incaser Informatica S.L. - Sergio Teruel
+# (c) 2015 Incaser Informatica S.L. - Carlos Dauden
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+{
+    "name": "Portal Welcome Email Template",
+    "summary": "Use an email template for send portal invitation",
```
Comment (elicoidal) [inline portal_welcome_email_template/__openerp__.py]: Adds an customizable email template for portal user invitation

File: portal_welcome_email_template/README.rst (hunk 11 above)
Comment (pedrobaeza) [inline portal_welcome_email_template/README.rst]: I would put "The sent email is enqueued instead of being directly processed".

**Comment by sergio-teruel** [comment]
@hbrunn @pedrobaeza @elicoidal Changes done. Thanks for review.

**Comment by hbrunn** [comment]
@sergio-incaser great, now only the test is missing

**Comment by sergio-teruel** [comment]
@hbrunn Tests added.

**Comment by hbrunn** [comment]
:+1:

**Comment by carlosdauden** [comment]
:+1:

**Comment by rafaelbn** [comment]
Thanks for contribution, very nice :+1:

**Pull request merged**